src/analysis/inner_product_space/finite_dimensional.lean

```diff
@@ -0,0 +1,412 @@
+/-
+Copyright (c) 2023 Monica Omar. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Monica Omar
+-/
+import analysis.inner_product_space.adjoint
+import analysis.inner_product_space.spectrum
+
+/-!
+# Finite-dimensional inner product spaces
+
+In this file, we prove some results in finite-dimensional inner product spaces.
+
+## Notation
+
+This file uses the local notation `P _` for `orthogonal_projection _`
+and `↥P _` for the extended orthogonal projection `V →L[ℂ] V`.
+
+We let `V` be an inner product space over `ℂ`.
+-/
+
+variables {V : Type*} [inner_product_space ℂ V]
+local notation `P` := orthogonal_projection
+
+/-- `U` is `T` invariant: `∀ u : V`, if `u ∈ U` then `T u ∈ U`-/
+def submodule.invariant (U : submodule ℂ V) (T : V →ₗ[ℂ] V) : Prop := U ≤ U.comap T
+lemma submodule.invariant_iff (U : submodule ℂ V) (T : V →ₗ[ℂ] V) :
+  submodule.invariant U T ↔ T '' U ⊆ U :=
+by simp only [set.image_subset_iff]; refl
+
+/-- `↥P _` is the extended version of `P _` -/
+noncomputable def orthogonal_projection.extend [finite_dimensional ℂ V]
+  (U : submodule ℂ V) : V →L[ℂ] V := U.subtypeL.comp (P U)
+local notation `↥P` := orthogonal_projection.extend
+
+lemma orthogonal_projection.extend_iff [finite_dimensional ℂ V] (U : submodule ℂ V) (x : V) :
+  ↥P U x = ↑(P U x) := rfl
+
+-- the extended orthogonal projection is an invariant subspace
+lemma submodule.invariant_of_ortho_proj (U : submodule ℂ V) [finite_dimensional ℂ V] :
+  submodule.invariant U (↥P U) :=
+begin
+  intros x hx,
+  rw [submodule.mem_comap, continuous_linear_map.coe_coe,
+      orthogonal_projection.extend_iff U, orthogonal_projection_eq_self_iff.mpr hx],
+  exact hx,
+end
```

Comment:
As a hint that this statement might not be interesting, it doesn't actually use any facts about orthogonal projections:
```suggestion
begin
  rw [submodule.invariant, orthogonal_projection.extend, continuous_linear_map.coe_comp,
    submodule.comap_comp, submodule.coe_subtypeL, submodule.comap_subtype_self,
    submodule.comap_top],
  exact le_top,
end
```

Reply:
That proofs reads simply as "`orthogonal_projection : V →L[ℂ] U` so by definition it lands in U."

Reply:
Yeah, you're right. But we might need to use the fact that it is invariant at some point?